rp2040-hal/src/intrinsics.rs

```diff
@@ -0,0 +1,244 @@
+/// Generate a series of aliases for an intrinsic function.
+macro_rules! intrinsics_aliases {
+    (
+        extern $abi:tt fn $name:ident( $($argname:ident: $ty:ty),* ) -> $ret:ty,
+    ) => {};
+    (
+        unsafe extern $abi:tt fn $name:ident( $($argname:ident: $ty:ty),* ) -> $ret:ty,
+    ) => {};
+
+    (
+        extern $abi:tt fn $name:ident( $($argname:ident: $ty:ty),* ) -> $ret:ty,
+        $alias:ident
+        $($rest:ident)*
+    ) => {
+        intrinsics! {
+            extern $abi fn $alias( $($argname: $ty),* ) -> $ret {
+                $name($($argname),*)
+            }
+        }
+
+        intrinsics_aliases! {
+            extern $abi fn $name( $($argname: $ty),* ) -> $ret,
+            $($rest)*
+        }
+    };
+
+    (
+        unsafe extern $abi:tt fn $name:ident( $($argname:ident: $ty:ty),* ) -> $ret:ty,
+        $alias:ident
+        $($rest:ident)*
+    ) => {
+        intrinsics! {
+            unsafe extern $abi fn $alias( $($argname: $ty),* ) -> $ret {
+                $name($($argname),*)
+            }
+        }
+
+        intrinsics_aliases! {
+            unsafe extern $abi fn $name( $($argname: $ty),* ) -> $ret,
+            $($rest)*
+        }
+    };
+}
+
+/// The macro used to define overridden intrinsics.
+///
+/// This is heavily inspired by the macro used by compiler-builtins.  The idea
+/// is to abstract anything special that needs to be done to override an
+/// intrinsic function.
+///
+/// Like the compiler-builtins macro, it accepts a series of functions that
+/// looks like normal Rust code:
+///
+///     intrinsics! {
+///         extern "C" fn foo(a: i32) -> u32 {
+///             // ...
+///         }
+///
+///         #[nonstandard_attribute]
+///         extern "C" fn bar(a: i32) -> u32 {
+///             // ...
+///         }
+///     }
+///
+/// Each function can also be decorated with nonstandard attributes to control
+/// additional behaviour:
+///
+/// * `slower_than_default` - indicates that the override is slower than the
+///   default implementation.  Currently this just disables the override
+///   entirely.
+/// * `bootrom_v2` - indicates that the override is only available
+///   on a V2 bootrom or higher.  Currently this is ignored and the
+///   override is always used.
```

Comment:
What happens on chips with V1 bootrom?
If it causes a crash or undefined behavior, I'd say that this is the wrong default.

Reply:
I don't know, I don't have a chip with a V1 bootrom to test.  I suspect it would crash: the table resolve would probably return 0, so you'd probably get a hard fault exception (which would panic).  However, I'm not actually sure how common V1 bootroms are (i.e. where they ever actually public?).  The Pico SDK [has logic](https://github.com/raspberrypi/pico-sdk/blob/2062372d203b372849d573f252cf7c6dc2800c0a/src/rp2_common/pico_double/double_init_rom.c#L26) that I'm not clear about: it only inserts the fallbacks if both V1 ROM and B0 chip support is enabled.

If they actually exist in the wild, then yeah I'd agree.

Reply:
As I understand, V1 ROM is equivalent to B0 chip. (B0 has ROM v1, B1 has ROM v2, B2 has ROM v3)
I didn't find any hard numbers, but after some searching I'd say that the original RP PICO boards sold in February 2021 were based on RP2040 revision B0, so there are probably quite a few in the wild.
When I'm back home I'll check which revisions I have.

Reply:
Well, probably best to make non-default then, so updated with them gated by the feature `rom-v2-intrinsics`.

Reply:
Just had a look, the first rp pico boards I bought really contained b0 stepping chips. From a local shop, no special early samples.

Reply:
And I confirm the expected behavior, more or less:
Building a firmware with `rom-v2-intrinsics` enabled, and flashing it to a RP2040 B0, causes a crash when executing a division.
What actually happens is not even a jump to address 0. Instead, the lookup for the `soft_double_table` returns 0, and then the `ddiv` method is looked up at an offset from that position. Incidently, the offset of the `ddiv` method is `0x0c`, and the memory location `0x0c` happens to contain a pointer to the HardFault handler in ROM, which is just a `loop { wfi() }`. Therefore, the debugger doesn't even show a crash.

I wonder if it would be better to somehow detect that situation and actively call panic!(), instead of just running into some random undefined behavior?

BTW, also tried on a RP2040 B1, works as expected.

Reply:
Well one idea might be to extend the `::ptr()` method of all ROM functions that require a V2 bootrom to check the bootrom version and panic if they're called at all on a V1.  The check is pretty simple: [just fetching the ROM at 0x13](https://github.com/raspberrypi/pico-sdk/blob/2062372d203b372849d573f252cf7c6dc2800c0a/src/rp2_common/pico_platform/include/pico/platform.h#L336) and comparing the integer number.  I'd probably just add an attribute for the ROM functions definitions and check it in a modified lookup body, and you could also add it to the general case for `soft_double_table()`.

The version might be a good idea to expose in general as part of the ROM functions.  Probably orthogonal to this PR, though, since it's a general problem with the ROM.

Reply:
Well, I added a check for a V1 bootrom in the lookup for the floating point functions.  So you should get a more sensible panic if you try and use them on an unsupported version, either explicitly or implicitly through an intrinsic.

Reply:
Great, works (panics) as expected:
```
INFO  Program start
└─ rp2040_project_template::__cortex_m_rt_main @ src/main.rs:34
ERROR panicked at 'Double precision operations require V2 bootrom (found: V1)', /home/jan/rp2040/rp-rs/rp-hal/rp2040-hal/src/rom_data.rs:384:9
└─ panic_probe::print_defmt::print @ /home/jan/.cargo/registry/src/github.com-1ecc6299db9ec823/panic-probe-0.3.0/src/lib.rs:91
```